designs/2022-forbid-disable/README.md

```diff
@@ -0,0 +1,120 @@
+- Repo: eslint/eslint
+- Start Date: 2022-02-28
+- RFC PR:
+- Authors: Brian Bartels
+
+# ESLintRC Option To Forbid Disable For Specific Rules
+
+## Summary
+
+Allow users to forbid users from disabling certain rules. This is similar to `noInlineConfig` where the primary difference is that `noInlineConfig` is set for all rules.
+
+## Motivation
+
+When working with complex lint configs. Some rules are more important than other. For this reason it can be inconsequential to disable certain rules. But incredibly bad to disable others.
+
+EG:
+```
+//eslint-disable-next-line @government/required-SECURITY-rule
+```
+Is a much worse rule to disable than
+```
+//eslint-disable-next-line noExtraSemi
+```
+
+`noInlineConfig` works well for some. But in reality, many users disable rules at least somewhat frequently for valid reasons. And in large repos, there are going to be exceptions that need inline configuration. 
+
+## Detailed Design
+
+in `linter.js` I will adjust the `getDirectiveComments` function so that I reference a list of forbidden rules. If someone is disabling a rule that is forbidden. I will add an error similar to `warnInlineConfig`
+
+In `eslintrc` I will add another option for rules. The new state can be
+```
+"warn"
+"error"
+"forbid"
+```
+where forbid is always an error and cannot be overridden by a directive
+
+I will add a new array called `forbiddenRules` that will be processed on initialization of the linter.js. This will happen in the Constructor. Then in `getDirectiveComments` I will reference the rule being disabled in the directive. If that rule matches a rule in `forbiddenRules` then I will not accept that directive. And I will declare an error on that string
```

Comment:
is this just things that are forbidden to have inline comments? do they only apply to disabling, or also changing between warn and error, or also changing configuration?

what about setting this in the root, so that a nested `.eslintrc` is prevented from overriding it?

can shared configs specify this list? Can it be specified in overrides?

Reply:
I think that leaving it the way it is is better. Imagine linting a mono-repo where one org controls one folder, and another controls another. One group might have different standards about what rules are forbid and which ones are error/warn. 

As for the first question. A rule set to forbid cannot be modified by a directive in any way.

The array `forbiddenRules` is a code array in linter.js, not set in eslintrc

Reply:
ok so it's not really "forbidden rules", it's more "rules that are not overrideable inline" (which isn't a great name either)

Reply:
Well forbid is an extension of error. So a rule that is forbidden will always error (not be configurable by directive). Pzixel gave me that name as it is used in Rust Linting.

Reply:
that's not how it reads to me; a forbidden rule is a rule nobody is allowed to use.

Reply:
what do you think about the name "strictError"

Reply:
If another config file can change it to not be an error, then that wouldn't be true either. If the only thing being restricted is inline configuration, i think that phrase needs to be in the term.

Reply:
The intention of this change is not to restrict the ability of developers from adjusting rules from configuration files. so a rule might be "strictError" in one path, and "error" in another (depending on the eslintrc). I feel like that is fairly intuitive when you consider the rest of the ESLint configuration.

I feel like the name "ErrorNoDirective" is less intuitive. That's my 2c

Reply:
I do think “forbid” is probably not the right word for this setting, maybe “enforce”?

Reply:
Well if I move forward with Nicholas's idea of extending noInlineConfig to accept an array of rules. Then This discussion will be null and void

Reply:
forbid has been forbidden. I changed the design quite a bit to reflect that change